### DISCUSSION TITLE
Sorting null values last using a custom sorter

### DISCUSSION THREAD

**Discussion by krivulcik** [ravendb/ravendb#19629] [General]
We have used the approach with the extra index field suggested here to have null values last in our queries: https://github.com/ravendb/ravendb/discussions/16703#discussioncomment-6250097

We have tried to implement a custom sorter to achieve this functionality, which would avoid the need to have extra fields in the index.

The proof of concept below works well for doubles (not tested thoroughly, also not deployed in production yet).

It also performs better than sorting using the extra field, with query times comparable to the standard sorter.

1) Corax currently doesn't support custom sorters. Is this something that can change in the future, i.e., will Corax receive support for custom sorters? Or there are limitations that prevent this?

2) Would it be possible to incorporate similar behavior to the "native" sorters in RavenDB? (I realize that it could be much more complex than it sounds.)

3) Any feedback on the custom sorter would be appreciated. It might contain some inefficiences, and also could be incorrect - as I mentioned, we didn't use it for anything serious yet.

``` csharp
using System;
using System.Collections.Generic;
using Lucene.Net.Index;
using Lucene.Net.Search;
using Lucene.Net.Store;

namespace HHRaven.Tryouts
{
    public sealed class CustomDoubleSorter : FieldComparator
    {
        private bool _reversed;
        private double[] _values;
        private double[] _currentReaderValues;
        private string _field;
        private DoubleParser _parser;
        private double _bottom;
        private static DoubleParser _ascSorter = new AnonymousClassDoubleParser();
        private static DoubleParser _descSorter = new AnonymousClassDescDoubleParser();

        public CustomDoubleSorter(string field, int numHits, int sortPos, bool reversed, List<string> diagnostics)
        {
            _values = new double[numHits];
            _reversed = reversed;
            _field = field;
            _parser = reversed ? _descSorter : _ascSorter;
        }

        public override int Compare(int slot1, int slot2)
        {
            double v1 = _values[slot1];
            double v2 = _values[slot2];
            if (v1 > v2)
            {
                return 1;
            }
            else if (v1 < v2)
            {
                return -1;
            }
            else
            {
                return 0;
            }
        }

        public override int CompareBottom(int doc, IState state)
        {
            double v2 = _currentReaderValues[doc];
            if (_bottom > v2)
            {
                return 1;
            }
            else if (_bottom < v2)
            {
                return -1;
            }
            else
            {
                return 0;
            }
        }

        public override void Copy(int slot, int doc, IState state)
        {
            _values[slot] = _currentReaderValues[doc];
        }

        public override void SetNextReader(IndexReader reader, int docBase, IState state)
        {
            _currentReaderValues = FieldCache_Fields.DEFAULT.GetDoubles(reader, _field, _parser, state);
        }

        public override void SetBottom(int bottom)
        {
            _bottom = _values[bottom];
        }

        public override IComparable this[int slot]
        {
            get { return (double)_values[slot]; }
        }
    }

    class AnonymousClassDoubleParser : DoubleParser
    {
        public virtual double ParseDouble(string value_Renamed)
        {
            if (value_Renamed == "NULL_VALUE") return double.MaxValue;

            return double.Parse(value_Renamed);
        }
        protected internal virtual object ReadResolve()
        {
            return FieldCache_Fields.DEFAULT_DOUBLE_PARSER;
        }
        public override string ToString()
        {
            return typeof(FieldCache).FullName + ".DEFAULT_DOUBLE_PARSER";
        }
    }

    class AnonymousClassDescDoubleParser : DoubleParser
    {
        public virtual double ParseDouble(string value_Renamed)
        {
            if (value_Renamed == "NULL_VALUE") return double.MinValue;
            return double.Parse(value_Renamed);
        }
        protected internal virtual object ReadResolve()
        {
            return FieldCache_Fields.DEFAULT_DOUBLE_PARSER;
        }
        public override string ToString()
        {
            return typeof(FieldCache).FullName + ".DEFAULT_DOUBLE_PARSER";
        }
    }
}
```

**Comment by ayende** [comment]
We don't plan on building custom sorting to Corax.
For reference, this query was a _challenge_ to build:

```
from Posts
where Category in ($categories)
order by PublishedDate desc
limit 15
```

Why is that a challenge, because you may have 1M records matching the query, and you need to sort all of them.
So Corax does a *lot* behind the scenes to _not_ have to read all the data and then sort it.

For that reason, to ensure that you can run this query _fast_, regardless of data size, we ensure that we always sort on sortable values, no custom can work and still be fast enough.

Having said that, You can also use this in the index: `MyField = doc.Field ?? double.MaxValue` and get it sorted last, or `MinValue` to get it sorted first.

**Reply by krivulcik** [reply]
Thanks for the information.

We need to have null values at the end when sorting both in ascending order (`null` is `MaxValue`) and descending order (`null` is `MinValue`). So storing one of the values won't work in both cases, and additional logic is required. That's why we wanted to explore the custom sorter route, and we'll need to evaluate whether using it is worth it (inability to switch to Corax in these cases).

**Reply by ayende** [reply]
Having said that, we actually hold the null values _separately_ from the rest of the values.
So we could probably implement something like:

`order by Name desc nulls first` and `order by Name asc nulls last`, etc.

**Reply by ayende** [reply]
That is a big feature, thoug

**Reply by krivulcik** [reply]
Thanks for the insight. I'm aware that there is more to it than it seems on the surface, so we aren't holding our breath for this feature. It would be great to have this functionality out of the box, though. Not sure how many projects need something like this - it sounds like a pretty reasonable requirement, but not many requests regarding this were in the discussions (or in the Google Groups forum).

**Reply by ayende** [reply]
The option with the `order by IsNull, Value` is the best option for now.
